### PR TITLE
feat: Added Resnet31 feature extractor (recognition)

### PR DESCRIPTION
**Pull request by charlesmindee**
This PR implements Resnet31 feature extractor as described in [(https://arxiv.org/pdf/1811.00751.pdf)]

### CODE REVIEW

File: doctr/models/resnet.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+
+class ResnetBlock(layers.Layer):
+    """Implements a resnet31 block with shortcut
+
+    Args:
+        conv_shortcut: Use of shortcut
+        output_channels: number of channels to use in Conv2D
+        kernel_size: size of square kernels
+
```
Comment (fg-mindee) [inline doctr/models/resnet.py:21]: You can remove the blank line here

File: doctr/models/resnet.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+            )
+        else:
+            self.shortcut = layers.Lambda(lambda x: x)
+        self.conv_block = Sequential(
+            self.conv_resnetblock(output_channels, 3)
+        )
+        self.act = layers.Activation('relu')
+
+    @staticmethod
+    def conv_resnetblock(output_channels, kernel_size):
```
Comment (fg-mindee) [inline doctr/models/resnet.py:52]: Could you add some typing here please?

File: doctr/models/resnet.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        return [
+            layers.Conv2D(output_channels, kernel_size, padding='same', use_bias=False, kernel_initializer='he_normal'),
+            layers.BatchNormalization(),
+            layers.Activation('relu'),
+            layers.Conv2D(output_channels, kernel_size, padding='same', use_bias=False, kernel_initializer='he_normal'),
+            layers.BatchNormalization(),
```
Comment (fg-mindee) [inline doctr/models/resnet.py:58]: Could you use `conv_sequence` instead please? (kernel initializer, bias, and padding are already set correctly by default)

File: doctr/models/resnet.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        return out
+
+
+class ResnetStage(Sequential):
+    """Implements a resnet31 stage
+
+    Args:
+        num_blocks: number of blocks inside the stage
+        output_channels: number of channels to use in Conv2D
+
```
Comment (fg-mindee) [inline doctr/models/resnet.py:78]: Could you remove this blank line please?

File: doctr/models/resnet.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+class Resnet31(Sequential):
+    """Resnet31 architecture with rectangular pooling windows as described in
+    `"Show, Attend and Read:A Simple and Strong Baseline for Irregular Text Recognition",
+    <https://arxiv.org/pdf/1811.00751.pdf>`_. Downsizing: (H, W) --> (H/8, W/4)
+
+    Args:
+        input_size: size of the images
+
```
Comment (fg-mindee) [inline doctr/models/resnet.py:102]: Same here

File: doctr/models/resnet.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+        ]
+        super().__init__(_layers)
+
+    @staticmethod
+    def conv_bn_act(output_channels, kernel_size, **kwargs):
+        return [
+            layers.Conv2D(output_channels, kernel_size, padding='same', kernel_initializer='he_normal', **kwargs),
+            layers.BatchNormalization(),
+            layers.Activation('relu'),
+        ]
```
Comment (fg-mindee) [inline doctr/models/resnet.py:129]: You can use `conv_sequence` from `doctr.models.utils` which is achieving the same thing :ok_hand:

File: doctr/models/resnet.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+
+    @staticmethod
+    def conv_bn_act_pool(output_channels, kernel_size, p_size, **kwargs):
+        return [
+            layers.Conv2D(output_channels, kernel_size, padding='same', kernel_initializer='he_normal', **kwargs),
+            layers.BatchNormalization(),
+            layers.Activation('relu'),
+            layers.MaxPool2D(pool_size=p_size, strides=p_size, padding='valid'),
+        ]
```
Comment (fg-mindee) [inline doctr/models/resnet.py:138]: To avoid duplication, I suggest combining `conv_sequence` and add the MaxPool2D after it in the constructor

File: doctr/models/resnet.py [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+        super().__init__()
+        final_blocks = [
+            ResnetBlock(output_channels, conv_shortcut=False) for _ in range(1, num_blocks)
+        ]
+        self.add(ResnetBlock(output_channels, conv_shortcut=True))
+        for final_block in final_blocks:
+            self.add(final_block)
+
+
+class Resnet31(Sequential):
```
Comment (fg-mindee) [inline doctr/models/resnet.py:95]: Could you rename this to `ResNet` and add a function `resnet31` that will instantiate the config of resnet31 ?
In the same fashion as in https://github.com/mindee/doctr/blob/main/doctr/models/vgg.py

File: test/test_models.py [hunk 9]
```diff
@@ -313,3 +313,12 @@ def test_zoo_models(arch_name):
...
     assert isinstance(model, models.OCRPredictor)
+
+
+def test_resnet31():
+    resnet_model = models.Resnet31(input_size=(32, 128, 3))
+    resnet_input = tf.random.uniform(shape=[8, 32, 128, 3], minval=0, maxval=1)
+    resnet_out = resnet_model(inputs=resnet_input)
+    assert isinstance(resnet_out, tf.Tensor)
+    assert isinstance(resnet_model, tf.keras.Model)
+    assert resnet_out.numpy().shape == (8, 4, 32, 512)
```
Comment (fg-mindee) [inline test/test_models.py:324]: You can remove this, and only add a line right after L264
```
["resnet31", False, (32, 128, 3), (4, 32, 512)],
```
from what I understand

**Review by fg-mindee — CHANGES_REQUESTED** [review]
Thanks for the PR! I added many simplification suggestions, let me know what you think!

**Review by fg-mindee — APPROVED** [review]
Thanks, this looks quite clean now!
